### PR TITLE
Android support

### PR DESCRIPTION
**Pull request by JBou**
This PR builds and packs a .aar library for Android that contains native libraries for the currently [supported ABIs](https://developer.android.com/ndk/guides/abis#gradle). An .aar library is like a .jar library, specifically for Android, that also supports native-libraries. More on the aar file structure [here](https://developer.android.com/studio/projects/android-library.html#aar-contents).

I also added a GitHub workflow as example to build the library. A sample workflow run and the resulting .aar library attached as artifact can be found [here](https://github.com/JBou/netty-incubator-codec-quic/actions/runs/695594000).

The minimum Android API level is 21. It's tied to the minAPI of quiche.
Note that on API 21-23 you have to enable [Java 8+ API desugaring support](https://developer.android.com/studio/write/java8-support#library-desugaring), otherwise the app will crash, it won't find a lambda class. On Android API Level >=24, Java 8 features used in netty-incubator-codec-quic (like lambdas) are supported and it's working without enabling desugaring.

I made a sample application [here](https://github.com/JBou/netty-quic-demo) to test this out. It includes the sample server and client code from the [example package](https://github.com/netty/netty-incubator-codec-quic/tree/main/src/test/java/io/netty/incubator/codec/quic/example) and a simple Android client.
~~Currently, when sending messages from the Android client to the server works perfectly fine, but when receiving messages from the server on the Android client the app crashes with a native exception. The GitHub issue can be found [here](https://github.com/netty/netty-incubator-codec-quic/issues/243). It needs to be fixed to make this implementation work, but currently I don't know if the problem is caused by netty or quiche or some android specific problem. I would kindly ask your help to solve it, because I'm new to C/native programming and still had a lot to learn and fix to make this library cross-compile for Android.~~

I changed the library name on Android to netty_quiche.so, because most of the times an Android phone supports multiple architectures and falls back to other architectures, so the name needs to be the same. Secondly, I needed to change it because the PlatformDependent class returned `netty_quiche_linux_aarch_64` on Android.

We should check if the `extraCflags`, `extraCxxflags`, `cmakeAsmFlags`, `cmakeCFlags`, `cmakeCxxFlags`, `cflags`, `extraLdflags` are correct. I'm not quite familiar with the flags.

I'm looking forward to fix this crash and make this PR work. I'm sure there are still some things to change before merging this, just let me know. 
Making `QUIC` available as transport protocol for Netty on Android would be a great step forward, especially because it has some great features that are useful on mobile networks and mobile devices (such as [connection migration](https://tools.ietf.org/id/draft-tan-quic-connection-migration-00.html#name-connection-migration), ...). Thanks for all of your hard work on this framework and I hope I can contribute something back to the community with this addition.

### CODE REVIEW

File: src/main/c/netty_quic_boringssl.c [hunk 1]
```diff
@@ -787,7 +787,7 @@ jint netty_boringssl_JNI_OnLoad(JNIEnv* env, const char* packagePrefix) {
 
 
     NETTY_JNI_UTIL_LOAD_CLASS(env, byteArrayClass, "[B", done);
-    NETTY_JNI_UTIL_LOAD_CLASS(env, stringClass, "Ljava/lang/String;", done);
+    NETTY_JNI_UTIL_LOAD_CLASS(env, stringClass, "java/lang/String", done);
```
Comment (normanmaurer) [inline src/main/c/netty_quic_boringssl.c:790]: why this change ?

File: src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java [hunk 2]
```diff
@@ -27,9 +27,9 @@
 final class BoringSSLCertificateVerifyCallback {
 
     private final QuicheQuicSslEngineMap engineMap;
-    private final X509ExtendedTrustManager manager;
+    private final X509TrustManager manager;
 
-    BoringSSLCertificateVerifyCallback(QuicheQuicSslEngineMap engineMap, X509ExtendedTrustManager manager) {
+    BoringSSLCertificateVerifyCallback(QuicheQuicSslEngineMap engineMap, X509TrustManager manager) {
```
Comment (normanmaurer) [inline src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: Why this change ? This doesn't look correct as we want to pass the engine down to the manager

File: src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java [hunk 3]
```diff
@@ -117,13 +117,13 @@ private X509ExtendedKeyManager chooseKeyManager(KeyManagerFactory keyManagerFact
         throw new IllegalArgumentException("No X509ExtendedKeyManager included");
     }
 
-    private static X509ExtendedTrustManager chooseTrustManager(TrustManagerFactory trustManagerFactory) {
+    private static X509TrustManager chooseTrustManager(TrustManagerFactory trustManagerFactory) {
```
Comment (normanmaurer) [inline src/main/java/io/netty/incubator/codec/quic/QuicheQuicSslContext.java:120]: same comment as above... I think we want to use the extended manager

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: src/main/c/netty_quic_boringssl.c (hunk 1 above)
Comment (JBou) [reply src/main/c/netty_quic_boringssl.c:790]: I had to change this because otherwise it was throwing an exception and crashed, there is more information on this page:
https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class

It seems that on the desktop it still works without `L` and `;`, The unit tests are all successful.

File: src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java (hunk 2 above)
Comment (JBou) [reply src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: The minSdkVersion of quiche is Android API 21. `X509ExtendedTrustManager` is only available on API 24+, as you can see [in the javadocs in the upper right corner](https://developer.android.com/reference/javax/net/ssl/X509ExtendedTrustManager). Increasing the minSdkVersion from 21 to 24 drops support for 20% of the devices, so I would really like to keep it at 21, especially because quiche also keeps the api at 21, I would try to avoid dropping support for 20% of the devices just for a single class.

I found [this PR](https://github.com/netty/netty/pull/5912/files) and saw that the methods we call in `BoringSSLCertificateVerifyCallback`, namely `manager.checkServerTrusted(peerCerts, authAlgorithm, engine);` and `manager.checkClientTrusted(peerCerts, authAlgorithm, engine);` are delegated to the X509TrustManager, so i just replaced the `X509ExtendedTrustManager` with `X509TrustManager` and though it was okay. But maybe I'm wrong, as you said.
Would it help to wrap the trustmanager, when possible? like seen on these lines: https://github.com/netty/netty/blob/master/handler/src/main/java/io/netty/handler/ssl/util/SimpleTrustManagerFactory.java#L139-L141
As the declarations are not public and we just call methods of the X509TrustManager, that class could be used and just wrapped on platforms where it is available, or is that a bad idea?

Comment (JBou) [reply src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: These are commits from other projects with the same problem:

https://github.com/eclipse/californium/pull/1593

https://github.com/google/conscrypt/pull/645

https://github.com/BiglySoftware/BiglyBT/commit/2e0a2f9d7e20223f3ded89b5c161883abd3f978c

Comment (normanmaurer) [reply src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: https://github.com/netty/netty-incubator-codec-quic/pull/249 can you check this one...

File: src/main/c/netty_quic_boringssl.c (hunk 1 above)
Comment (normanmaurer) [reply src/main/c/netty_quic_boringssl.c:790]: ok fair enough...  Thanks a lot!

File: src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java (hunk 2 above)
Comment (JBou) [reply src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: I will test it tomorrow evening on an old device

Comment (normanmaurer) [reply src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: Thanks a lot!

Comment (JBou) [reply src/main/java/io/netty/incubator/codec/quic/BoringSSLCertificateVerifyCallback.java:32]: It works! But I tried it with the change I requested on the PR (`Class.forName(X509ExtendedTrustManager.class.getName());`).

Just wanted to note that on API 21-23 I had to enable [Java 8+ API desugaring support](https://developer.android.com/studio/write/java8-support#library-desugaring), otherwise the app will crash, it won't find a lambda class. On Android API Level >=24, Java 8 features used in netty-incubator-codec-quic (like lambdas) are supported and it's working without enabling desugaring.

**Comment by normanmaurer** [comment]
@JBou can you please rebase on latest main branch ?

**Comment by JBou** [comment]
Yes, I will do it tonight 👍 Have you seen the changes I requested on #249?

**Comment by normanmaurer** [comment]
@JBou no there are no comments on the PR... hmm. Can you open a PR with the requested changes ?

**Comment by JBou** [comment]
Oh, now that's strange. The reviews are still there but are only being shown when I'm logged in. It might be that I left the browser-tab open over night and there is a bug with GitHub and my session. Really strange 🤔
Nevertheless, I will open a PR right now with the requested changes.

image

**Comment by normanmaurer** [comment]
@JBou ok lets rebase now

**Comment by JBou** [comment]
I rebased yesterday

File: src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="@JAVA_MODULE_NAME@" android:versionCode="1" android:versionName="1.0">
```
Comment (normanmaurer) [inline src/main/AndroidManifest.xml:1]: please add the license header (just like its done in the pom.xml

File: src/main/native-package/m4/custom.m4.android.template [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+dnl  You may obtain a copy of the License at
+dnl  
+dnl     http://www.apache.org/licenses/LICENSE-2.0
+dnl  
+dnl  Unless required by applicable law or agreed to in writing, software
+dnl  distributed under the License is distributed on an "AS IS" BASIS,
+dnl  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+dnl  See the License for the specific language governing permissions and
+dnl  limitations under the License.
+dnl ---------------------------------------------------------------------------
```
Comment (normanmaurer) [inline src/main/native-package/m4/custom.m4.android.template:16]: you should also add the netty license header

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: src/main/AndroidManifest.xml (hunk 4 above)
Comment (JBou) [reply src/main/AndroidManifest.xml:16]: done

File: src/main/native-package/m4/custom.m4.android.template (hunk 5 above)
Comment (JBou) [reply src/main/native-package/m4/custom.m4.android.template]: done

**Comment by JBou** [comment]
@normanmaurer do you have some time to review the rest of this PR?

**Comment by normanmaurer** [comment]
@JBou yep I will do it this week.

**Comment by normanmaurer** [comment]
@JBou sorry but I need to ask you one last time to rebase :/

**Comment by JBou** [comment]
I rebased again 👍

**Comment by normanmaurer** [comment]
@JBou thanks a lot!

**Comment by JBou** [comment]
Awesome, thanks for merging!
